python/valis/cache.py

```diff
@@ -0,0 +1,288 @@
+from fastapi_cache.backends.inmemory import InMemoryBackend
+from fastapi_cache.backends.redis import RedisBackend
+from fastapi_cache.decorator import _augment_signature, _locate_param
+from redis.asyncio.client import Redis
+from starlette.requests import Request
+from starlette.responses import Response
+from starlette.status import HTTP_304_NOT_MODIFIED
+
+from valis.settings import settings
+
+
+if TYPE_CHECKING:
+    from typing import AsyncIterator
+
+    from fastapi import FastAPI
+    from fastapi_cache.coder import Coder
+    from fastapi_cache.types import KeyBuilder
+
+
+__all__ = ['valis_cache', 'lifespan', 'valis_cache_key_builder']
+
+
+P = ParamSpec("P")
+R = TypeVar("R")
+
+
+logger = logging.getLogger("uvicorn.error")
+
+CACHE_TTL: float = 15_552_000 # 6 months
+
+
+@asynccontextmanager
+async def lifespan(_: FastAPI) -> AsyncIterator[None]:
+    backend = settings.cache_backend
+    if backend == 'in-memory':
+        logger.info('Using in-memory backend for caching')
+        FastAPICache.init(InMemoryBackend(),
+                          prefix="fastapi-cache",
+                          key_builder=valis_cache_key_builder)
+    elif backend == 'redis':
+        logger.info('Using Redis backend for caching')
+        redis = Redis.from_url("redis://localhost")
+        FastAPICache.init(RedisBackend(redis),
+                          prefix="fastapi-cache",
+                          key_builder=valis_cache_key_builder)
+    elif backend == 'null' or not backend:
+        logger.info('Using null backend for caching')
+        FastAPICache.init(NullCacheBackend(),
+                          prefix="fastapi-cache",
+                          key_builder=valis_cache_key_builder)
+    else:
+        raise ValueError(f'Invalid cache backend {backend}')
+
+    yield
+
+
+async def valis_cache_key_builder(
+    func,
+    namespace: str = "",
+    request: Request | None = None,
+    _: Response | None = None,
+    *args,
+    **kwargs,
+):
+    query_params = request.query_params.items() if request else []
+
+    try:
+        body_json = await request.json()
+        body = sorted(body_json.items()) if body_json else []
+    except json.JSONDecodeError:
+        body = []
+
+    hash = hashlib.new('md5')
+    for param,value in list(query_params) + body:
+        hash.update(param.encode())
+        hash.update(str(value).encode())
+
+    params_hash = hash.hexdigest()[0:8]
+
+    url = request.url.path.replace('/', '_') if request else ""
+    if url.startswith('_'):
+        url = url[1:]
+
+    chunks = [
+            namespace,
+            request.method.lower() if request else "",
+            url,
+            params_hash,
+        ]
+
+    return ":".join(chunks)
+
+
+def valis_cache(
+    expire: Optional[int] = CACHE_TTL,
+    coder: Optional[Type[Coder]] = None,
+    key_builder: Optional[KeyBuilder] = None,
+    namespace: str = "valis-cache",
+    injected_dependency_namespace: str = "__fastapi_cache",
```

Comment:
can you add some comments to this function at the lines where you modified the code from the original function?

Reply:
Done